src/lib.rs

```diff
@@ -0,0 +1,564 @@
+    max_parallel: u32,
+    stop_flag: &AtomicBool,
+    max_behind: u32, /* maximum number of seconds a node can be behind before it is deemed
+                      * "behind" */
+    hooks: &mut H,
+) -> Result<(), NodeError>
+where
+    H: NodeHooks<D>, {
+    // Use TLS if the URI scheme is HTTPS.
+    // This uses whatever system certificates have been installed as trusted roots.
+    let node_ep = if node_ep.uri().scheme().map_or(false, |x| x == &http::uri::Scheme::HTTPS) {
+        node_ep.tls_config(ClientTlsConfig::new()).map_err(NodeError::ConnectionError)?
+    } else {
+        node_ep
+    };
+
+    let mut node = v2::Client::new(node_ep).await.map_err(NodeError::ConnectionError)?;
+    while !stop_flag.load(Ordering::Acquire) {
+        hooks.on_use_node(&mut node).await?;
+
+        let timeout = std::time::Duration::from_secs(max_behind.into());
+        let mut finalized_blocks = node.get_finalized_blocks_from(*height).await?;
+
+        let (has_error, chunks) = finalized_blocks
+            .next_chunk_timeout(max_parallel as usize, timeout)
+            .await
+            .map_err(|_| NodeError::Timeout)?;
+
+        for fb in chunks {
+            let data = hooks.on_finalized_block(&mut node, &fb).await?;
+
+            if sender.send(data).await.is_err() {
+                log::error!("The database connection has been closed. Terminating node queries.");
+                return Ok(());
+            }
+
+            *height = height.next();
+        }
+
+        if has_error {
+            // we have processed the blocks we can, but further queries on the same stream
+            // will fail since the stream signalled an error.
+            return Err(NodeError::OtherError(anyhow::anyhow!("Finalized block stream dropped.")));
+        }
+    }
+
+    Ok(())
+}
+
+/// Executes service infrastructure. Handles connections to a set of nodes and a
+/// database (configured with `app_config`) on their
+/// own separate threads.
+/// Implementation of `NH` defines hooks used by the node thread, while
+/// implementation of `DH` defines hooks used by the database thread.
+pub async fn run_service<D, P, DH, NH>(
+    sql_schema: &'static str,
+    app_config: SharedIndexerArgs,
+    mut node_hooks: NH,
+    config_logger: Option<impl FnOnce(&mut env_logger::Builder, log::LevelFilter)>,
+) -> Result<(), anyhow::Error>
+where
+    P: PrepareStatements + Send + 'static,
+    D: Send + Sync + 'static,
+    DH: DatabaseHooks<D, P> + Send + Sync + 'static,
+    NH: NodeHooks<D>, {
+    anyhow::ensure!(!app_config.endpoint.is_empty(), "At least one node must be provided.");
+    let db_config = app_config.config;
+
+    let mut log_builder = env_logger::Builder::from_env("TRANSACTION_LOGGER_LOG");
+    log_builder.filter_module(module_path!(), app_config.log_level);
+    if let Some(config_logger) = config_logger {
+        config_logger(&mut log_builder, app_config.log_level);
+    }
+    log_builder.init();
```

Comment:
This should be left up to the user and not be in the library

Reply:
Yes, I agree. Will move it to the consumer end.